Zooming the map by scrolling does not work as expected on Touchpad

Hi,

I have a touchpad that uses two finger inputs to scroll. Scrolling works in every other program. In Minutor, attempting to scroll will either do nothing, zoom *in* to a random zoom level, or zoom *out* to a random zoom level (including randomly jerking in and out across multiple zoom levels even if I am scrolling continuously on one direction on the touchpad). 

This is on KDE Plasma (latest released version) on Linux. I believe my touchpad is using the xf86-input-synaptics driver.

Things I have tried:

 * Set up a script with `xdotool` to send one "scroll up" input every five seconds. This *works*. Minutor goes up one zoom level every five seconds.

 * Checked the inputs being sent to programs by the mouse driver with `xev`. If I scroll extremely slowly, I can get the touchpad to only send two scroll events at a time. I can't get it down to one. Note that scrolling extremely slowly does not seem to solve the problem with scrolling in Minutor.

If you don't immediately know how to fix this problem, or fixing it would create other issues, I think it would be nice to have some kind of keyboard input for scrolling as a around, and possible a zoom slider on the right of the screen as well, like the "layers" slider.

I'm compiling it, the one I was using when I opened this was https://github.com/mrkite/minutor/commit/57a47ff which is from the end of March. I'm trying to update to make sure the issue wasn't fixed, but I can't compile it at present. Bisecting reveals that https://github.com/mrkite/minutor/commit/c456e1d75dfc60a951d72d107616ec1ebffb6b62 introduced the problem. (Yes, I cherrypicked https://github.com/mrkite/minutor/commit/5d5a65ee753dcc182bce38338a42086915a891b1 to avoid that compilation issue.)

    /usr/bin/ld: properties.o: in function `Properties::DisplayProperties(QVariant)':
    properties.cpp:(.text+0xe8): undefined reference to `void PropertieTreeCreator::ParseIterable<QHash<QString, QVariant> >(QTreeWidgetItem*, QHash<QString, QVariant> const&)'
    /usr/bin/ld: properties.cpp:(.text+0x1e7): undefined reference to `void PropertieTreeCreator::ParseIterable<QMap<QString, QVariant> >(QTreeWidgetItem*, QMap<QString, QVariant> const&)'

I'm on QT version 5.15.0. I can copy this into a new issue if you want.

I think it's worth noting that most of the time Minutor doesn't seem to react to my scroll events at all. The bug would be more understandable to me if it always zoomed in more quickly than I wanted because it was reacting to 2 scroll events instead of just one, but most of the time it seems to ignore them.

--- a/mapview.cpp
+++ b/mapview.cpp
@@ -2,6 +2,7 @@
 #include <QPainter>
 #include <QResizeEvent>
 #include <QMessageBox>
+#include <cmath>
 #include <assert.h>
 
 #include "mapview.h"
@@ -140,7 +141,7 @@ void MapView::clearCache() {
 void MapView::adjustZoom(double steps, bool allowZoomOut)
 {
   // get current zoom level as an index, rounded to nearest int
-  int oldZoomIndex = zoomLevel + 0.5;
+  int oldZoomIndex = (int)(floor(zoomLevel + 0.5));
   zoomLevel += steps;
 
   // use Fibonacci numbers to get natural zoom behaviour
@@ -151,7 +152,7 @@ void MapView::adjustZoom(double steps, bool allowZoomOut)
   if (zoomLevel < zoomMin) zoomLevel = zoomMin;
   if (zoomLevel > zoomMax) zoomLevel = zoomMax;
 
-  int zoomIndex = zoomLevel + 0.5;
+  int zoomIndex = (int)(floor(zoomLevel + 0.5 ));
 
   // check whether zoomIndex has changed since last adjustZoom call
   // don't return early if steps == 0, this is used for initialization

--- a/mapview.cpp
+++ b/mapview.cpp
@@ -16,7 +16,7 @@ MapView::MapView(QWidget *parent)
   : QWidget(parent)
   , depth(255)
   , scale(1)      // overworld coordinate mapping
-  , zoomIndex(0)  // 1:1
+  , zoomLevel(0)  // 1:1
   , cache(ChunkCache::Instance())
 {
   adjustZoom(0, false);
@@ -139,15 +139,23 @@ void MapView::clearCache() {
 
 void MapView::adjustZoom(double steps, bool allowZoomOut)
 {
-  zoomIndex += steps;
+  // get current zoom level as an index, rounded to nearest int
+  int oldZoomIndex = zoomLevel + 0.5;
+  zoomLevel += steps;
 
   // use Fibonacci numbers to get natural zoom behaviour
   const float zoomTable[] = {1, 2, 3, 5, 8, 13, 21, 34, 55, 89};
   const int zoomMin = allowZoomOut ? -4 : 0;
   const int zoomMax = (sizeof(zoomTable) / sizeof(float)) -1;
 
-  if (zoomIndex < zoomMin) zoomIndex = zoomMin;
-  if (zoomIndex > zoomMax) zoomIndex = zoomMax;
+  if (zoomLevel < zoomMin) zoomLevel = zoomMin;
+  if (zoomLevel > zoomMax) zoomLevel = zoomMax;
+
+  int zoomIndex = zoomLevel + 0.5;
+
+  // check whether zoomIndex has changed since last adjustZoom call
+  // don't return early if steps == 0, this is used for initialization
+  if (zoomIndex == oldZoomIndex && steps != 0) return;
 
   // determine minimal zoomed value that is allowed
   // with current window size and available pyhiscal memory
@@ -174,6 +182,9 @@ void MapView::adjustZoom(double steps, bool allowZoomOut)
 
   // we try to set higher margin than above (100%)!
   cache.setCacheMaxSize(2.0 * chunks);
+
+  // no point in redrawing if the zoom level didn't change
+  if (steps != 0) redraw();
 }
 
 static int lastMouseX = -1, lastMouseY = -1;
@@ -248,12 +259,10 @@ void MapView::wheelEvent(QWheelEvent *event) {
     emit demandDepthChange(event->delta() / 120);
   } else if ((event->modifiers() & modifier4ZoomOut) == modifier4ZoomOut) {
     // allow change zoom also to zoom OUT
-    adjustZoom( event->delta() / 120.0, true );
-    redraw();
+    adjustZoom( event->angleDelta().y() / 120.0, true );
   } else {
     // normal change zoom
-    adjustZoom( event->delta() / 120.0, false );
-    redraw();
+    adjustZoom( event->angleDelta().y() / 120.0, false );
   }
 }
 
@@ -579,7 +588,7 @@ void MapView::getToolTip(int x, int z) {
   hovertext += " [Cache:"
             + QString().number(this->cache.getCacheUsage()) + "/"
             + QString().number(this->cache.getCacheMax()) + "]";
-  hovertext += " Zoom:" + QString().number(zoomIndex);
+  hovertext += " Zoom:" + QString().number(zoomLevel);
 #endif
 
   emit hoverTextChanged(hovertext);

--- a/mapview.h
+++ b/mapview.h
@@ -104,7 +104,7 @@ class MapView : public QWidget {
   int depth;
   double x, z;
   int scale;
-  int zoomIndex;
+  double zoomLevel;
   double zoom;
   int flags;
   ChunkCache &cache;

